### PR TITLE
RFAI-03: Proposal provenance and outcomes ledger foundation

### PR DESCRIPTION
**Pull request by Chris0Jeky**
## Summary

Foundational slice of #975 (RFAI-03: Proposal generator V1 with verified provenance and outcomes ledger). Builds the domain types, application interfaces, infrastructure persistence, and comprehensive test coverage needed before wiring the full IProposalGenerator.

- **Domain types**: ProposalProvenance, ProvenanceField (extractive/inferred with confidence), EvidenceLink (structured source references), FieldVerificationResult (value object), ProposalOutcome (content-free decision ledger), plus ProvenanceKind/VerificationStatus/OutcomeDecision enums
- **Application layer**: IFuzzyTextMatcher with Levenshtein sliding-window implementation, IProposalOutcomeRepository, IDeterministicPreExtractor with ExtractedEntity type
- **Infrastructure**: DeterministicPreExtractor using Microsoft.Recognizers.Text for dates/numbers/durations/URLs/emails, ProposalOutcomeConfiguration, ProposalOutcomeRepository, EF Core migration, DI registration
- **Tests**: 130 new tests (77 domain + 52 application + 1 API) covering all new types, validation rules, edge cases, fuzzy matching, and malformed input resilience

### Key design decisions

- **ProposalOutcome is content-free**: only stores structural dimensions (decision type, latency, field counts, risk level) -- never proposal text or user content (GP-10 compliance)
- **FuzzyTextMatcher** uses normalized case-insensitive Levenshtein distance with whitespace collapsing and sliding window for best substring similarity
- **DeterministicPreExtractor** handles malformed input by catching exceptions per-recognizer and returning partial results
- **ProvenanceField** enforces that extractive fields must have an extractive quote, and confidence stays in [0.0, 1.0]

### Note

Full IProposalGenerator wiring awaits #974 merge. This PR is self-contained and can be merged independently.

## Test plan

- [x] All 130 new tests pass (77 domain entity tests + 52 application service tests + 1 architecture test)
- [x] Full test suite passes: 5,196 tests, 0 failures (6 pre-existing skips)
- [x] `dotnet build backend/Taskdeck.sln -c Release` succeeds with no new warnings
- [x] EF Core migration generates cleanly
- [x] ProposalOutcome entity verified to contain no PII-capable properties

Closes: part of #975

### CODE REVIEW

**Comment by Chris0Jeky** [comment]
## Adversarial Self-Review

### Issues Found and Fixed

1. **FuzzyTextMatcher: Negative similarity possible** -- When the Levenshtein edit distance exceeded the candidate length (e.g., comparing short candidate against very different text), the similarity formula `1.0 - distance / length` could return a negative value. **Fixed** by clamping to `Math.Max(0.0, similarity)`.

2. **DeterministicPreExtractor: Redundant DateTimeRecognizer call** -- `DateTimeRecognizer.RecognizeDateTime` was called twice (once for dates skipping durations, once for durations only). This doubled the recognition cost. **Fixed** by processing both date and duration results in a single call, branching on `TypeName.Contains("duration")`.

### Verified OK

3. **ProposalOutcome content-free (no PII)** -- Confirmed: no property stores user-generated text. Only structural dimensions (decision enum, latency, field counts, risk/source labels as short strings, model ID). A structural test `Constructor_ContentFree_ShouldNotStoreAnyUserContent` verifies no property named Summary/Description/Content/Text/Body exists.

4. **Architecture layer boundaries** -- FuzzyTextMatcher (pure logic, no infra deps) is correctly in Application layer. DeterministicPreExtractor (depends on Microsoft.Recognizers.Text NuGet) is correctly in Infrastructure. Domain entities have no infrastructure references.

5. **Recognizers.Text malformed input** -- Each recognizer call is wrapped in try/catch, returning partial results on failure. Tested with null, empty, whitespace-only, random garbage, unicode, and very long inputs.

6. **Fuzzy matching edge cases** -- Tested: exact match, exact substring, close match (typo), no match, empty/null strings, whitespace-only, unicode, case insensitivity, whitespace normalization, single character, candidate longer than source, very long source text. Threshold boundary at 0.0 tested.

7. **ProvenanceField extractive quote enforcement** -- Extractive kind requires a non-empty quote; inferred kind allows null quote. Both paths tested.

8. **EvidenceLink span validation** -- SpanEnd < SpanStart is rejected. Negative spans rejected. SpanStart == SpanEnd accepted (zero-length span). Only SpanStart without SpanEnd accepted.

### Remaining Notes

- The FuzzyTextMatcher sliding window is O(n * m * k) which is fine for typical extractive quote verification (~50-200 char quotes against ~500-2000 char sources). For very large documents, a more efficient algorithm (e.g., Smith-Waterman or suffix array) could be warranted -- but that's a follow-up optimization, not a correctness issue.
- Full IProposalGenerator wiring awaits #974 merge.

**Comment by Chris0Jeky** [comment]
## Adversarial Review (Round 2)

### Bot Findings Addressed

**Codex P2: FieldVerificationResult.cs:63 -- Enforce verification-status/confidence consistency**
Fixed. The constructor now validates that `Status` and `AdjustedConfidence` are consistent with the documented contract:
- `Verified` -> `AdjustedConfidence == OriginalConfidence`
- `Downgraded` -> `AdjustedConfidence < OriginalConfidence`
- `Failed` -> `AdjustedConfidence == 0.0`
- `Unverified` -> no constraint (not yet processed)

Added 7 new test cases covering all valid and invalid combinations. Commit: `62e17410`.

**Codex P2: DeterministicPreExtractor.cs:81 -- Avoid running DateTime recognition twice**
Already resolved in this PR version. The code uses a single `DateTimeRecognizer.RecognizeDateTime` call and splits results by `TypeName.Contains("duration")`. No fix needed.

### Additional Findings Fixed

**ProposalProvenance.AddField -- No parent-ID validation (domain invariant gap)**
`AddField` accepted any `ProvenanceField` regardless of its `ProposalProvenanceId`, allowing silent mis-parenting. Fixed: now validates `field.ProposalProvenanceId == Id` and adds a null guard. 2 new tests. Commit: `bdb5b00b`.

**ProposalOutcomeRepository -- No upper bound on limit parameter**
`GetByUserIdAsync` and `GetByDecisionAsync` accepted arbitrary `limit` values with no ceiling, meaning a caller could request `int.MaxValue` rows. Fixed: clamped to `MaxLimit = 1000`. Commit: `3d3252dc`.

### Findings Noted (Not Fixed -- Design Observations)

**FuzzyTextMatcher -- O(n * m * k) sliding-window complexity**
The sliding window iterates over multiple window sizes and positions, each computing a full Levenshtein distance. For large source texts (thousands of chars) this could become slow. Current usage appears bounded to capture-sized inputs, so this is acceptable for now, but if this is ever used in batch processing or on large documents, consider adding an input-length guard or switching to an O(n*m) semi-global alignment algorithm.

**Application.Tests references Infrastructure project**
`DeterministicPreExtractorTests` lives in `Taskdeck.Application.Tests` but instantiates `Taskdeck.Infrastructure.Services.DeterministicPreExtractor` directly, requiring a project reference from Application.Tests to Infrastructure. This violates the test-layer-mirroring convention (the test should live in an Infrastructure.Tests or Integration.Tests project). Low priority, but worth noting for future cleanup.

### Verification

- `dotnet build backend/Taskdeck.sln -c Release` -- 0 errors
- 138 targeted tests pass (85 Domain + 52 Application + 1 Api)
- 20 architecture tests pass (16 passed, 4 skipped -- pre-existing skips)

File: backend/src/Taskdeck.Domain/Entities/ProposalOutcome.cs [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+        double? averageFieldConfidence = null)
+    {
+        if (proposalId == Guid.Empty)
+            throw new DomainException(ErrorCodes.ValidationError, "ProposalId cannot be empty");
+        if (decidedByUserId == Guid.Empty)
+            throw new DomainException(ErrorCodes.ValidationError, "DecidedByUserId cannot be empty");
+        if (!Enum.IsDefined(decision))
+            throw new DomainException(ErrorCodes.ValidationError, "OutcomeDecision value is invalid");
+        if (decisionLatencySeconds < 0)
+            throw new DomainException(ErrorCodes.ValidationError, "DecisionLatencySeconds cannot be negative");
```
[the remark this answers is not in the document]
Comment (Chris0Jeky) [reply backend/src/Taskdeck.Domain/Entities/ProposalOutcome.cs]: Fixed in `f7c9d7be`. `ProposalOutcome` now rejects all non-finite latency and average confidence values (`NaN`, `+Infinity`, `-Infinity`) before persistence. Added regression coverage in `ProposalOutcomeTests` and verified with the documented RFAI-03 backend filter plus full `Taskdeck.Domain.Tests`.

File: backend/src/Taskdeck.Domain/Entities/ProvenanceField.cs [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+        FieldName = fieldName;
+        Kind = kind;
+        Confidence = confidence;
+        ProposalProvenanceId = proposalProvenanceId;
+        ExtractiveQuote = extractiveQuote;
+    }
+
+    public void AddEvidenceLink(EvidenceLink link)
+    {
+        _evidenceLinks.Add(link);
```
[the remark this answers is not in the document]
Comment (Chris0Jeky) [reply backend/src/Taskdeck.Domain/Entities/ProvenanceField.cs]: Fixed in `f7c9d7be`. `ProvenanceField.AddEvidenceLink` now rejects `null` links and keeps the existing `ProvenanceFieldId == Id` ownership guard, with regression tests for both mismatched ownership and null input. Targeted provenance tests pass.

File: backend/src/Taskdeck.Domain/Entities/FieldVerificationResult.cs [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+        if (string.IsNullOrWhiteSpace(fieldName))
+            throw new DomainException(ErrorCodes.ValidationError, "FieldName cannot be empty");
+        if (!Enum.IsDefined(status))
+            throw new DomainException(ErrorCodes.ValidationError, "VerificationStatus value is invalid");
+        if (originalConfidence < 0.0 || originalConfidence > 1.0)
+            throw new DomainException(ErrorCodes.ValidationError, "OriginalConfidence must be between 0.0 and 1.0");
+        if (adjustedConfidence < 0.0 || adjustedConfidence > 1.0)
+            throw new DomainException(ErrorCodes.ValidationError, "AdjustedConfidence must be between 0.0 and 1.0");
+        if (similarityScore.HasValue && (similarityScore.Value < 0.0 || similarityScore.Value > 1.0))
+            throw new DomainException(ErrorCodes.ValidationError, "SimilarityScore must be between 0.0 and 1.0");
```
[the remark this answers is not in the document]
Comment (Chris0Jeky) [reply backend/src/Taskdeck.Domain/Entities/FieldVerificationResult.cs]: Fixed in `f7c9d7be`. `FieldVerificationResult` now rejects non-finite `OriginalConfidence`, `AdjustedConfidence`, and `SimilarityScore` values before status-consistency checks run. Added focused tests for `NaN`, `+Infinity`, and `-Infinity` on each value.

File: backend/src/Taskdeck.Domain/Entities/ProvenanceField.cs [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+        string? extractiveQuote = null)
+    {
+        if (string.IsNullOrWhiteSpace(fieldName))
+            throw new DomainException(ErrorCodes.ValidationError, "FieldName cannot be empty");
+        if (fieldName.Length > 100)
+            throw new DomainException(ErrorCodes.ValidationError, "FieldName cannot exceed 100 characters");
+        if (!Enum.IsDefined(kind))
+            throw new DomainException(ErrorCodes.ValidationError, "ProvenanceKind value is invalid");
+        if (confidence < 0.0 || confidence > 1.0)
+            throw new DomainException(ErrorCodes.ValidationError, "Confidence must be between 0.0 and 1.0");
```
[the remark this answers is not in the document]
Comment (Chris0Jeky) [reply backend/src/Taskdeck.Domain/Entities/ProvenanceField.cs]: Fixed in `f7c9d7be`. `ProvenanceField` constructor and `DowngradeConfidence` now reject non-finite confidence values. Added regression coverage for `NaN`, `+Infinity`, and `-Infinity` in `ProvenanceFieldTests`.

File: backend/src/Taskdeck.Domain/Entities/ProposalOutcome.cs [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+        if (string.IsNullOrWhiteSpace(riskLevel))
+            throw new DomainException(ErrorCodes.ValidationError, "RiskLevel cannot be empty");
+        if (riskLevel.Length > 50)
+            throw new DomainException(ErrorCodes.ValidationError, "RiskLevel cannot exceed 50 characters");
+        if (modelId != null && modelId.Length > 100)
+            throw new DomainException(ErrorCodes.ValidationError, "ModelId cannot exceed 100 characters");
+        if (averageFieldConfidence.HasValue && (double.IsNaN(averageFieldConfidence.Value) || averageFieldConfidence.Value < 0.0 || averageFieldConfidence.Value > 1.0))
+            throw new DomainException(ErrorCodes.ValidationError, "AverageFieldConfidence must be between 0.0 and 1.0");
+        if (decision != OutcomeDecision.EditedThenApproved && editedFieldCount > 0)
+            throw new DomainException(ErrorCodes.ValidationError, "EditedFieldCount must be 0 when decision is not EditedThenApproved");
```
[the remark this answers is not in the document]
Comment (Chris0Jeky) [reply backend/src/Taskdeck.Domain/Entities/ProposalOutcome.cs:91]: Fixed in `f7c9d7be`. `ProposalOutcome` now requires `EditedThenApproved` outcomes to have `EditedFieldCount > 0`, while still rejecting edits on non-edit decisions. Added a direct regression test for the zero-edit contradiction.

File: backend/src/Taskdeck.Infrastructure/Services/DeterministicPreExtractor.cs [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+/// Extracts deterministic structured data (dates, numbers, durations, URLs, emails)
+/// from free text using Microsoft.Recognizers.Text.
+///
+/// This runs entirely locally -- no LLM calls, no network access.
+/// Handles malformed input gracefully by returning empty results.
+/// </summary>
+public class DeterministicPreExtractor : IDeterministicPreExtractor
+{
+    private static readonly Regex UrlRegex = new(
+        @"https?://[^\s<>""')\]]+",
```
[the remark this answers is not in the document]
Comment (Chris0Jeky) [reply backend/src/Taskdeck.Infrastructure/Services/DeterministicPreExtractor.cs]: Already fixed by `2ae51bf2`, which trims/excludes trailing terminal punctuation from URL extraction. The documented RFAI-03 provenance filter was rerun after the latest fixes and passes.

File: backend/src/Taskdeck.Domain/Entities/FieldVerificationResult.cs [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
```
[the remark this answers is not in the document]
Comment (Chris0Jeky) [reply backend/src/Taskdeck.Domain/Entities/FieldVerificationResult.cs]: Already fixed earlier on this branch by `62e17410`. The latest pass keeps those status/confidence consistency guards and adds non-finite confidence rejection in `f7c9d7be`; the provenance/domain filters pass.

File: backend/src/Taskdeck.Infrastructure/Services/DeterministicPreExtractor.cs [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+        }
+        catch
+        {
+            // Malformed input -- skip number extraction
+        }
+
+        // Duration recognition (uses DateTimeRecognizer with duration type)
+        try
+        {
+            var durationResults = DateTimeRecognizer.RecognizeDateTime(text, culture);
```
[the remark this answers is not in the document]
Comment (Chris0Jeky) [reply backend/src/Taskdeck.Infrastructure/Services/DeterministicPreExtractor.cs]: Already fixed earlier on this branch by `4138770a`; `Extract` performs a single `DateTimeRecognizer.RecognizeDateTime` pass and splits date/duration results from that collection. The RFAI-03 provenance filter was rerun after the latest fix commit and passes.

**Comment by Chris0Jeky** [comment]
## Review-fix update

Pushed `f7c9d7be` (`Harden proposal provenance invariants`) to `rfai-03/proposal-provenance-v2`.

### Fixed
- `ProposalOutcome` now rejects non-finite decision latency and average field confidence values (`NaN`, `+Infinity`, `-Infinity`).
- `ProposalOutcome` now rejects `EditedThenApproved` ledger rows with zero edited fields.
- `FieldVerificationResult` now rejects non-finite original/adjusted/similarity confidence values before status consistency checks.
- `ProvenanceField` now rejects non-finite confidence values in construction/downgrade paths and null evidence links in `AddEvidenceLink`, while preserving evidence ownership validation.

### Verification
- `dotnet test backend/tests/Taskdeck.Domain.Tests/Taskdeck.Domain.Tests.csproj -c Release --filter "FullyQualifiedName~ProposalOutcome|FullyQualifiedName~ProvenanceField|FullyQualifiedName~FieldVerificationResult"` -- passed, 84 tests.
- `dotnet test backend/Taskdeck.sln -c Release -m:1 --filter "FullyQualifiedName~Provenance|FullyQualifiedName~FieldVerification|FullyQualifiedName~ProposalOutcome|FullyQualifiedName~FuzzyText|FullyQualifiedName~PreExtractor"` -- passed, 97 Domain + 87 Application + 8 API tests.
- `dotnet test backend/tests/Taskdeck.Domain.Tests/Taskdeck.Domain.Tests.csproj -c Release` -- passed, 1072 tests.
- `gh pr checks 993 --repo Chris0Jeky/Taskdeck` -- all active required checks passing; PR merge state is `CLEAN`.

I replied to the active Codex inline threads with the commit/test evidence. No canonical docs were changed for this narrow invariant hardening.

File: backend/src/Taskdeck.Domain/Entities/ProvenanceField.cs [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+        if (!Enum.IsDefined(kind))
+            throw new DomainException(ErrorCodes.ValidationError, "ProvenanceKind value is invalid");
+        if (!double.IsFinite(confidence) || confidence < 0.0 || confidence > 1.0)
+            throw new DomainException(ErrorCodes.ValidationError, "Confidence must be between 0.0 and 1.0");
+        if (proposalProvenanceId == Guid.Empty)
+            throw new DomainException(ErrorCodes.ValidationError, "ProposalProvenanceId cannot be empty");
+        if (kind == ProvenanceKind.Extractive && string.IsNullOrWhiteSpace(extractiveQuote))
+            throw new DomainException(ErrorCodes.ValidationError, "ExtractiveQuote is required for Extractive provenance kind");
+        if (extractiveQuote != null && extractiveQuote.Length > 2000)
+            throw new DomainException(ErrorCodes.ValidationError, "ExtractiveQuote cannot exceed 2000 characters");
```
[the remark this answers is not in the document]
Comment (Chris0Jeky) [reply backend/src/Taskdeck.Domain/Entities/ProvenanceField.cs:68]: Fixed in `9b1586ba`. `ProvenanceField` now rejects non-null `ExtractiveQuote` values unless the kind is `Extractive`, and `ProvenanceFieldTests` covers the inferred-field contradiction. Targeted Domain tests passed.

File: backend/src/Taskdeck.Infrastructure/Services/DeterministicPreExtractor.cs [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+/// Extracts deterministic structured data (dates, numbers, durations, URLs, emails)
+/// from free text using Microsoft.Recognizers.Text.
+///
+/// This runs entirely locally -- no LLM calls, no network access.
+/// Handles malformed input gracefully by returning empty results.
+/// </summary>
+public class DeterministicPreExtractor : IDeterministicPreExtractor
+{
+    private static readonly Regex UrlRegex = new(
+        @"https?://[^\s<>""')\]]+(?<=[^\s<>""')\].,;:!?])",
```
[the remark this answers is not in the document]
Comment (Chris0Jeky) [reply backend/src/Taskdeck.Infrastructure/Services/DeterministicPreExtractor.cs]: Fixed in `9b1586ba`. URL extraction now permits valid `)` and `]` inside URL bodies and trims only trailing sentence punctuation/unbalanced closing delimiters. Added deterministic extractor regression coverage for Wikipedia-style parentheses, IPv6 brackets, and sentence punctuation; targeted Application tests passed.

**Comment by Chris0Jeky** [comment]
Addressed the migration review finding in `551e797b`.

Changes:
- `ExtendProposalOutcomesForMetrics` now backfills `Decision` from legacy `OutcomeType` before relying on the non-null column value.
- Added migration bootstrap regression coverage that migrates to the legacy outcome schema, inserts all four legacy `OutcomeType` values, applies the metrics migration, and asserts matching `OutcomeDecision` values.

Verification:
- `dotnet test backend/tests/Taskdeck.Api.Tests/Taskdeck.Api.Tests.csproj -c Release --filter "FullyQualifiedName~MigrationBootstrap" --logger "console;verbosity=minimal"` passed: 6 tests.
- `dotnet test backend/tests/Taskdeck.Domain.Tests/Taskdeck.Domain.Tests.csproj -c Release --filter "FullyQualifiedName~ProposalOutcome" --logger "console;verbosity=minimal"` passed: 34 tests.

File: backend/src/Taskdeck.Infrastructure/Persistence/Migrations/20260425173300_ExtendProposalOutcomesForMetrics.cs [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+                type: "INTEGER",
+                nullable: false,
+                defaultValue: 0);
+
+            migrationBuilder.Sql("""
+                UPDATE ProposalOutcomes
+                SET Decision = CASE OutcomeType
+                    WHEN 0 THEN 0
+                    WHEN 1 THEN 1
+                    WHEN 2 THEN 2
```
[the remark this answers is not in the document]
Comment (Chris0Jeky) [reply backend/src/Taskdeck.Infrastructure/Persistence/Migrations/20260425173300_ExtendProposalOutcomesForMetrics.cs:35]: Fixed in 515af1de. The migration now includes a SQL backfill step after the `EditedFieldCount` and `FieldCount` columns are added:

```sql
UPDATE ProposalOutcomes
SET EditedFieldCount = 1,
    FieldCount = 1
WHERE OutcomeType = 1
```

This sets both columns to `1` as a non-zero sentinel for legacy `EditedThenApproved` rows (exact counts are not recoverable from the old schema). Non-edited outcome types keep their default `0/0`.

A new migration bootstrap test (`ExtendProposalOutcomesForMetrics_backfills_edit_counts_for_legacy_edited_rows`) verifies that:
- `EditedThenApproved` rows get `EditedFieldCount = 1, FieldCount = 1`
- `Approved`, `Rejected`, and `Ignored` rows keep `0/0`
